crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -0,0 +1,226 @@
+                let executed_sell_amount = executed
+                    .checked_mul(prices.buy)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(prices.sell)
+                    .ok_or(Error::DivisionByZero)?;
+                // Sell slightly more `sell_token` to capture the `surplus_fee`
+                let executed_sell_amount_with_fee = executed_sell_amount
+                    .checked_add(surplus_fee)
+                    .ok_or(Error::Overflow)?;
+                // Scale to support partially fillable orders
+                let limit_sell_amount = sell_amount
+                    .checked_mul(executed)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(buy_amount)
+                    .ok_or(Error::DivisionByZero)?;
+                // Remaining surplus after fees
+                // Do not return error if `checked_sub` fails because violated limit prices will
+                // be caught by simulation
+                let surplus = limit_sell_amount
+                    .checked_sub(executed_sell_amount_with_fee)
+                    .unwrap_or(eth::U256::zero());
+                Ok(surplus
+                    .checked_mul(eth::U256::from_f64_lossy(factor * 100.))
+                    .ok_or(Error::Overflow)?
+                    / 100)
+            }
+            Side::Sell => {
+                // How much `buy_token` we get for `executed` amount of `sell_token`
+                let executed_buy_amount = executed
+                    .checked_mul(prices.sell)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(prices.buy)
+                    .ok_or(Error::DivisionByZero)?;
+                let executed_sell_amount_with_fee =
+                    executed.checked_add(surplus_fee).ok_or(Error::Overflow)?;
+                // Scale to support partially fillable orders
+                let limit_buy_amount = buy_amount
+                    .checked_mul(executed_sell_amount_with_fee)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(sell_amount)
+                    .ok_or(Error::DivisionByZero)?;
+                // Remaining surplus after fees
+                // Do not return error if `checked_sub` fails because violated limit prices will
+                // be caught by simulation
+                let surplus = executed_buy_amount
+                    .checked_sub(limit_buy_amount)
+                    .unwrap_or(eth::U256::zero());
+                let surplus_in_sell_token = surplus
+                    .checked_mul(prices.buy)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(prices.sell)
+                    .ok_or(Error::DivisionByZero)?;
+                Ok(surplus_in_sell_token
+                    .checked_mul(eth::U256::from_f64_lossy(factor * 100.))
+                    .ok_or(Error::Overflow)?
+                    / 100)
```

Comment:
nit: Might be slightly nicer to read if you only return `surplus_in_sell_token` from both `match` arms and then apply the `factor` at the end of the function because this is identical for both cases.
Same comment applies to `volume_fee()`.

Reply:
extracted into separate function. also refactored the big match into smaller parts - focus more on the values that are calculated.